lib/src/package.dart

```diff
@@ -202,120 +205,71 @@ class Package {
-            : '$beneath/$file';
-      }).expand((file) {
-        if (fileExists(file)) return [file];
-        if (!dirExists(file)) return [];
-
-        // `git ls-files` only returns files, except in the case of a submodule
-        // or a symlink to a directory.
-        return recursive ? _listWithinDir(file) : [file];
-      });
-    } else {
-      files = listDir(beneath,
-          recursive: recursive, includeDirs: false, allowed: _allowedFiles);
-    }
-
-    return files.where((file) {
-      // Using substring here is generally problematic in cases where dir has
-      // one or more trailing slashes. If you do listDir("foo"), you'll get back
-      // paths like "foo/bar". If you do listDir("foo/"), you'll get "foo/bar"
-      // (note the trailing slash was dropped. If you do listDir("foo//"),
-      // you'll get "foo//bar".
-      //
-      // This means if you strip off the prefix, the resulting string may have a
-      // leading separator (if the prefix did not have a trailing one) or it may
-      // not. However, since we are only using the results of that to call
-      // contains() on, the leading separator is harmless.
-      assert(file.startsWith(beneath));
-      file = file.substring(beneath.length);
-      return !_disallowedFiles.any(file.endsWith) &&
-          !_disallowedDirs.any(file.contains);
-    }).toList();
+    final visitedDirs = <String>{};
+    return Ignore.listFiles(
+      beneath: beneath,
+      listDir: (dir) {
+        var contents = Directory(resolve(dir)).listSync();
+        if (!recursive) {
+          contents = contents.where((entity) => entity is! Directory).toList();
+        }
+        return contents.map((entity) {
+          final relative = p.relative(entity.path, from: this.dir);
+          if (Platform.isWindows) {
+            return p.posix.joinAll(p.split(relative));
+          }
+          return relative;
+        });
+      },
+      ignoreForDir: (dir) {
+        final pubIgnore = resolve('$dir/.pubignore');
+        final gitIgnore = resolve('$dir/.gitignore');
+        final ignoreFile = fileExists(pubIgnore)
+            ? pubIgnore
+            : (fileExists(gitIgnore) ? gitIgnore : null);
+
+        final rules = [
+          if (dir == '.') ..._basicIgnoreRules,
+          if (ignoreFile != null) readTextFile(ignoreFile),
+        ];
+        return rules.isEmpty
+            ? null
+            : Ignore(
+                rules,
+                onInvalidPattern: (pattern, exception) {
+                  log.warning(
+                      '$ignoreFile had invalid pattern $pattern. ${exception.message}');
+                },
+              );
+      },
+      isDir: (dir) {
+        final resolved = resolve(dir);
+        final isDir = dirExists(resolved);
+        if (isDir) {
+          final canonicalized = resolveSymlinksOfDir(resolved);
+          // Avoid cycles by only visiting each location once.
+          return visitedDirs.add(canonicalized);
```

Comment:
If we say `isDir` false won't `createTarGz` try to add it to the to a tarball with `file.openRead()`?
https://github.com/dart-lang/pub/blob/11e7b2ce977cb27b4a508963ec1ace01a56bb6e1/lib/src/io.dart#L922-L940

Also won't this mean that in:
```
/folder/
  hello.txt
/linkA -> ../folder/
/linkB -> ../folder/
```
We'll get files as:
* `[/folder/hello.txt]`, OR,
* `[/linkA/hello.txt]`, OR,
* `[/linkB/hello.txt]`.

depending on which branch we visit first.

-------

Hmm, this is going to be rather complicated to support in any predictable manner.. Maybe we should disallow symlinks to folders?

It's easy to implement something where we only follow a symlink the first time we see it.. but the outcome from such a scan easily becomes rather complicated.

The only predictable options I see are:
 * (A) don't allow symlinks to directories,
 * (B) don't allow nested symlinks (even acyclic ones) -- simply throw an error if we see the same symlink more than once.
 * (C) don't allow symlink cycles -- throw an error if we see a symlink cycle.

(C) will allow a scenario like:
```
/folder/
  hello.txt
/linkA -> folder/
/linkB -> ../linkA/
/linkC -> ../linkA/
```
But then we need to track all the symlinks seen in the current stack, and throw an error if we see the same symlink twice.. and in this case we would then include the file `hello.txt` in the tarball 4 times!

So maybe (A) or (B) isn't so bad, though (B) seems like an arbitrary level of support for symlinks.

Reply:
You are right - it all becomes dodgy when we do this...

I would like to support 'simple' use-cases, such as sharing a directory between two packages in a mono-repo... but perhaps it is not worth the complexity.

Reply:
Forbidding directory symlinks for now (option A). We can add it later if the pressure grows for this.